File: 通用代码/状态模式/Context.java
//不成文的约束:
//把状体对象声明为静态常量,有几个状态对象就声明几个静态常量
//环境角色具有状态抽象角色定义的所有行为,具体执行使用委托方式
public class Context{
    //定义状态
    public final static State STATE1 = new ConcreteState1();
    public final static State STATE2 = new ConcreteState2();
    //当前状态
    private State currentState;
    //获得当前状态
    public State getCurrentState(){
        return currentState;
    }
    //设置当前状态
    public void setCurrentState(State _currentState){
        this.currentState = _currentState;
        //切换状态
        this.currentState.setContext(this);
    }
    //行为委托
    public void handle1(){
        this.currentState.handle1();
    }
    public void handle2(){
        this.currentState.handle2();
    }
}

File: 通用代码/状态模式(C++)/Context.cpp
#include "Context.h"

State* Context::pSTATE1 = NULL;
State* Context::pSTATE2 = NULL;

Context::Context(void)
{
    m_pCurrentState = NULL;
    pSTATE1 = new ConcreteState1();
    pSTATE2 = new ConcreteState2();
	cout<<"Context..."<<endl;
}

Context::~Context(void)
{
    delete pSTATE1;
    pSTATE1 = NULL;
    delete pSTATE2;
    pSTATE2 = NULL;
	cout<<"~Context..."<<endl;
}

State* Context::GetCurrentState()
{
    return m_pCurrentState;
}

void Context::SetCurrentState(State *pCurrentState)
{
    this->m_pCurrentState = pCurrentState;
	//切换状态
    this->m_pCurrentState->SetContext(this);
}

//行为委托
void Context::Handle1()
{
    this->m_pCurrentState->Handle1();
}

void Context::Handle2()
{
    this->m_pCurrentState->Handle2();
}
